improver/blending/spatial_weights.py

```diff
@@ -289,15 +150,150 @@ def create_template_slice(cube_to_collapse, blend_coord):
+            input_weights (iris.cube.Cube):
+                1D cube of shape self.blend_coord
+
+        Returns:
+            iris.cube.Cube:
+                Matching 3D shape of template cube, with values from 1D
+                weights cube
+        """
+        weights = iris.cube.CubeList()
+        for tslice, wslice in zip(
+            template.slices_over(self.blend_coord),
+            input_weights.slices_over(self.blend_coord),
+        ):
+            tslice.data[:, :] = wslice.data
+            weights.append(tslice)
+        weights = weights.merge_cube()
+        weights.data = weights.data.astype(FLOAT_DTYPE)
+        return weights
+
+    def _normalise_initial_weights(self, weights):
+        """Normalise weights so that they add up to 1 along the blend dimension
+        at each spatial point.  This is different from the normalisation that
+        happens after the application of fuzzy smoothing near mask boundaries.
+
+        Args:
+            weights (iris.cube.Cube)
+                3D weights containing zeros for masked points, but before
+                fuzzy smoothing
+
+        Returns:
+            iris.cube.Cube
+        """
+        weights_sum = np.sum(weights.data, axis=self.blend_axis)
+        normalised_weights = iris.cube.CubeList()
+        for weights_slice in weights.slices_over(self.blend_coord):
+            normalised_data = np.where(
+                weights_sum > 0, np.divide(weights_slice.data, weights_sum), 0
+            )
+            normalised_weights.append(
+                weights_slice.copy(data=normalised_data.astype(FLOAT_DTYPE))
+            )
+        weights = normalised_weights.merge_cube()
+        return weights
+
+    def _rescale_masked_weights(self, weights):
+        """Apply fuzzy smoothing to weights at the edge of masked areas
+
+        Args:
+            weights (iris.cube.Cube):
+                Pre-normalised weights where the weights of masked data points
+                have been set to 0
+
+        Returns:
+            (tuple): tuple containing:
+                **weights** (iris.cube.Cube):
+                    Weights where MASKED slices have been rescaled, but UNMASKED
+                    slices have not
+                **rescaled** (iris.cube.Cube):
+                    Binary (0/1) map showing which weights have been rescaled
+
+        """
+        is_rescaled = iris.cube.CubeList()
+        rescaled_weights = iris.cube.CubeList()
+        for weights_slice in weights.slices_over(self.blend_coord):
+            weights_nonzero = np.where(weights_slice.data > 0, 1, 0)
+            if np.all(weights_nonzero == 1):
+                # if there are no masked points in this slice, keep current weights
+                # and mark as unchanged (not rescaled)
+                rescaled_weights.append(weights_slice)
+                is_rescaled.append(
+                    weights_slice.copy(data=np.zeros(weights_slice.shape, dtype=int))
```

Comment:
This could be a `bool` dtype I think, as the only allowed values are 0 or 1.

Reply:
Done/